src/tasks/MonoTargetsTasks/MarshalingPInvokeScanner/MarshalingPInvokeScanner.cs

```diff
@@ -0,0 +1,333 @@
+
+    public override bool Execute()
+    {
+        if (Assemblies is null || Assemblies!.Length == 0)
+        {
+            Log.LogError($"{nameof(MarshalingPInvokeScanner)}.{nameof(Assemblies)} cannot be empty");
+            return false;
+        }
+
+        try
+        {
+            ExecuteInternal();
+            return !Log.HasLoggedErrors;
+        }
+        catch (LogAsErrorException e)
+        {
+            Log.LogError(e.Message);
+            return false;
+        }
+    }
+
+    private void ExecuteInternal()
+    {
+        if (Assemblies is not null)
+            IncompatibleAssemblies = ScanAssemblies(Assemblies);
+    }
+
+    private string[] ScanAssemblies(string[] assemblies)
+    {
+        HashSet<string> incompatible = new HashSet<string>();
+        MinimalMarshalingTypeCompatibilityProvider mmtcp =  new MinimalMarshalingTypeCompatibilityProvider();
+        foreach (string aname in assemblies)
+        {
+            if (IsAssemblyIncompatible(aname, mmtcp))
+                incompatible.Add(aname);
+        }
+
+        if (mmtcp.IsSecondPassNeeded)
+        {
+            foreach (string aname in assemblies)
+                ResolveInconclusiveTypes(incompatible, aname, mmtcp);
+        }
+
+        return incompatible.ToArray();
+    }
+
+    private static string GetMethodName(MetadataReader mr, MethodDefinition md)
+    {
+        return mr.GetString(md.Name);
+    }
+
+    private void ResolveInconclusiveTypes(HashSet<string> incompatible, string assyPath, MinimalMarshalingTypeCompatibilityProvider mmtcp)
+    {
+        string assyName = MetadataReader.GetAssemblyName(assyPath).Name!;
+        HashSet<string> inconclusiveTypes = mmtcp.GetInconclusiveTypesForAssembly(assyName);
+        if(inconclusiveTypes.Count == 0)
+            return;
+
+        using FileStream file = new FileStream(assyPath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
+        using PEReader peReader = new PEReader(file);
+        MetadataReader mdtReader = peReader.GetMetadataReader();
+
+        SignatureDecoder<Compatibility, object> decoder = new SignatureDecoder<Compatibility, object>(
+            mmtcp, mdtReader, null!);
+
+        foreach (TypeDefinitionHandle typeDefHandle in mdtReader.TypeDefinitions)
+        {
+            TypeDefinition typeDef = mdtReader.GetTypeDefinition(typeDefHandle);
+            string fullTypeName = string.Join(":", mdtReader.GetString(typeDef.Namespace), mdtReader.GetString(typeDef.Name));
+
+            // This is not perfect, but should work right for enums defined in other assemblies,
+            // which is the only case where we use Compatibility.Inconclusive.
+            if (inconclusiveTypes.Contains(fullTypeName) &&
+                mmtcp.GetTypeFromDefinition(mdtReader, typeDefHandle, 0) != Compatibility.Compatible)
+            {
+                Log.LogMessage(MessageImportance.Low, string.Format("Type {0} is marshaled and requires marshal-ilgen.", fullTypeName));
```

Comment:
Just curious - why `string.Format` and not an interpolated string?

Reply:
I just prefer them, they look more similar to the old-school `printf`.